### PR TITLE
Added Threshold Query Builder

### PR DESCRIPTION
**Pull request by ravit-db**

### CODE REVIEW

File: src/databricks/labs/remorph/reconcile/query_builder.py [hunk 1]
```diff
@@ -163,10 +151,127 @@ def _construct_hash_query(table_name: str, query_filter: str, hash_expr: str, ke
...
+        column_mapping = self.table_conf.list_to_dict(ColumnMapping, "source_name")
+        transformations_dict = self.table_conf.list_to_dict(Transformation, "column_name")
+
+        threshold_columns = set(threshold.column_name for threshold in self.table_conf.thresholds)
+        join_columns = set(self.table_conf.join_columns)
+
+        if self.table_conf.jdbc_reader_options and self.layer == "source":
+            partition_column = {self.table_conf.jdbc_reader_options.partition_column}
+        else:
+            partition_column = set()
```
Comment (ganeshdogiparthi-db) [inline src/databricks/labs/remorph/reconcile/query_builder.py:227]: These column's identification should be solely done in one function which we already have _get_column_list

File: src/databricks/labs/remorph/reconcile/connectors/data_source.py [hunk 2]
```diff
@@ -20,11 +20,18 @@ def __init__(self, source: str, spark: SparkSession, ws: WorkspaceClient, scope:
         self.scope = scope
 
     @abstractmethod
-    def read_data(self, schema_name: str, catalog_name: str, query: str, table_conf: Tables) -> DataFrame:
+    def read_data(
+        self, catalog_name: str, schema_name: str, query: str, jdbc_reader_options: JdbcReaderOptions
```
Comment (nfx) [inline src/databricks/labs/remorph/reconcile/connectors/data_source.py:24]:
```suggestion
        self, catalog: str, schema: str, query: str, options: ReaderOptions
```
As long as you are changing signature, make arg names shorter

File: src/databricks/labs/remorph/reconcile/connectors/data_source.py [hunk 3]
```diff
@@ -20,11 +20,18 @@ def __init__(self, source: str, spark: SparkSession, ws: WorkspaceClient, scope:
...
+    ) -> DataFrame:
         return NotImplemented
 
     @abstractmethod
-    def get_schema(self, table_name: str, schema_name: str, catalog_name: str) -> list[Schema]:
+    def get_schema(
+        self,
+        catalog_name: str,
```
Comment (nfx) [inline src/databricks/labs/remorph/reconcile/connectors/data_source.py:31]:
```suggestion
        catalog: str,
```

File: src/databricks/labs/remorph/reconcile/connectors/data_source.py [hunk 4]
```diff
@@ -20,11 +20,18 @@ def __init__(self, source: str, spark: SparkSession, ws: WorkspaceClient, scope:
...
+        schema_name: str,
```
Comment (nfx) [inline src/databricks/labs/remorph/reconcile/connectors/data_source.py:32]:
```suggestion
        schema: str,
```
Shorter

File: src/databricks/labs/remorph/reconcile/connectors/data_source.py [hunk 5]
```diff
@@ -20,11 +20,18 @@ def __init__(self, source: str, spark: SparkSession, ws: WorkspaceClient, scope:
...
+        table_name: str,
```
Comment (nfx) [inline src/databricks/labs/remorph/reconcile/connectors/data_source.py:33]:
```suggestion
        table: str,
```

File: src/databricks/labs/remorph/reconcile/query_builder.py [hunk 6]
```diff
@@ -8,129 +9,93 @@
...
-        # Remove threshold and drop columns
-        threshold_columns = {thresh.column_name for thresh in self.table_conf.thresholds or []}
-        if self.table_conf.drop_columns is None:
-            drop_columns = set()
-        else:
-            drop_columns = self._get_mapped_columns(self.layer, column_mapping, self.table_conf.drop_columns)
+    def _get_custom_transformation(self, columns, transformation_dict, column_mapping):
+        transformation_rule_mapping = []
+        for column in columns:
+            if column in transformation_dict.keys():
```
Comment (nfx) [inline src/databricks/labs/remorph/reconcile/query_builder.py:33]: Introduce a method that returns transformation rule mapping

File: src/databricks/labs/remorph/reconcile/query_builder.py [hunk 7]
```diff
@@ -8,129 +9,93 @@
...
-                transform_column = column
-
-            if transformations_dict and transform_column in transformations_dict.keys():
-                transformation = self._get_layer_transform(transformations_dict, transform_column, self.layer)
-            else:
-                column_data_type = schema.get(column).data_type
-                transformation = self._get_default_transformation(self.source, column_data_type).format(column)
+            column_origin = column if self.qc.layer == "source" else self._get_column_map(column, column_mapping)
+            column_data_type = schema.get(column_origin).data_type
+            transformation = self._get_default_transformation_expr(self.qc.db_type, column_data_type).format(
```
Comment (nfx) [inline src/databricks/labs/remorph/reconcile/query_builder.py:49]: Bug? Is there a format() method on this?

File: src/databricks/labs/remorph/reconcile/query_builder.py [hunk 8]
```diff
@@ -8,129 +9,93 @@
...
             )
-        if data_source == "databricks":
+        if data_source == SourceType.DATABRICKS.value:
             return DatabricksDataSource.databricks_datatype_mapper.get(
                 data_type, ColumnTransformationType.DATABRICKS_DEFAULT.value
             )
         msg = f"Unsupported source type --> {data_source}"
         raise ValueError(msg)
 
+    def _generate_transformation_rule_mapping(self, columns: list[str]) -> list[TransformRuleMapping]:
```
Comment (nfx) [inline src/databricks/labs/remorph/reconcile/query_builder.py:74]: It looks like this method can be reused up in the code

File: src/databricks/labs/remorph/reconcile/query_builder.py [hunk 9]
```diff
@@ -139,6 +104,48 @@ def _get_layer_transform(transform_dict: dict[str, Transformation], column: str,
     def _get_column_expr(func, column_transformations: list[TransformRuleMapping]):
         return [func(transformation) for transformation in column_transformations]
 
+    @staticmethod
+    def _get_column_map(column, column_mapping) -> str:
+        return column_mapping.get(column).target_name if column_mapping.get(column) else column
```
Comment (nfx) [inline src/databricks/labs/remorph/reconcile/query_builder.py:109]:
```suggestion
        return column_mapping.get(column, Column(target_name=column)).target_name
```
Use the second parameter of dict.get to return a default

File: src/databricks/labs/remorph/reconcile/query_builder.py [hunk 10]
```diff
@@ -162,11 +170,39 @@ def _construct_hash_query(table_name: str, query_filter: str, hash_expr: str, ke
...
     @staticmethod
-    def _get_mapped_columns(layer: str, column_mapping: dict, columns: list[str]) -> set[str]:
-        if layer == "source":
-            return set(columns)
-        select_columns = set()
-        for column in columns:
-            select_columns.add(column_mapping.get(column).target_name if column_mapping.get(column) else column)
-        return select_columns
+    def _construct_threshold_query(table_name, query_filter, threshold_columns_expr):
+        sql_query = StringIO()
```
Comment (nfx) [inline src/databricks/labs/remorph/reconcile/query_builder.py:198]: Simplify. Use f"x {y}"

File: src/databricks/labs/remorph/reconcile/query_config.py [hunk 11]
```diff
@@ -0,0 +1,71 @@
+from databricks.labs.remorph.reconcile.constants import SourceType
+from databricks.labs.remorph.reconcile.recon_config import (
+    ColumnMapping,
+    Schema,
+    Table,
+    Transformation,
+)
+
+
+class QueryConfig:  # pylint: disable=too-many-instance-attributes)
```
Comment (nfx) [inline src/databricks/labs/remorph/reconcile/query_config.py:10]:
```suggestion
class QueryConfig:
```
You are not allowed to disable linter warnings.

Convert instance attribute usage to methods.

**Review by nfx — CHANGES_REQUESTED** [review]

File: src/databricks/labs/remorph/reconcile/connectors/data_source.py (hunk 2 above)
Comment (ravit-db) [reply src/databricks/labs/remorph/reconcile/connectors/data_source.py]: fixed.

File: src/databricks/labs/remorph/reconcile/connectors/data_source.py (hunk 3 above)
Comment (ravit-db) [reply src/databricks/labs/remorph/reconcile/connectors/data_source.py]: fixed.

File: src/databricks/labs/remorph/reconcile/query_builder.py (hunk 1 above)
Comment (ravit-db) [reply src/databricks/labs/remorph/reconcile/query_builder.py]: fixed.

File: src/databricks/labs/remorph/reconcile/connectors/data_source.py (hunk 4 above)
Comment (ravit-db) [reply src/databricks/labs/remorph/reconcile/connectors/data_source.py]: fixed.

File: src/databricks/labs/remorph/reconcile/connectors/data_source.py (hunk 5 above)
Comment (ravit-db) [reply src/databricks/labs/remorph/reconcile/connectors/data_source.py]: fixed.

File: src/databricks/labs/remorph/reconcile/query_builder.py (hunk 7 above)
Comment (ravit-db) [reply src/databricks/labs/remorph/reconcile/query_builder.py:50]: it is  a str and it has a placeholder {} to use str.format

File: src/databricks/labs/remorph/reconcile/query_builder.py (hunk 8 above)
Comment (ravit-db) [reply src/databricks/labs/remorph/reconcile/query_builder.py:75]: actually _generate_transformation_rule_mapping () : # calls below two functions
                           _get_custom_transformation() # get the user defined transformations
                           _get_default_transformation() # get the default transformations

File: src/databricks/labs/remorph/reconcile/query_builder.py (hunk 9 above)
Comment (ravit-db) [reply src/databricks/labs/remorph/reconcile/query_builder.py]: fixed.

File: src/databricks/labs/remorph/reconcile/query_config.py (hunk 11 above)
Comment (ravit-db) [reply src/databricks/labs/remorph/reconcile/query_config.py]: fixed.

File: src/databricks/labs/remorph/reconcile/query_builder.py (hunk 10 above)
Comment (ravit-db) [reply src/databricks/labs/remorph/reconcile/query_builder.py]: fixed.

File: src/databricks/labs/remorph/reconcile/connectors/data_source.py [hunk 12]
```diff
@@ -20,11 +20,16 @@ def __init__(self, source: str, spark: SparkSession, ws: WorkspaceClient, scope:
         self.scope = scope
 
     @abstractmethod
-    def read_data(self, schema_name: str, catalog_name: str, query: str, table_conf: Tables) -> DataFrame:
+    def read_data(self, catalog: str, schema: str, query: str, jdbc_reader_options: JdbcReaderOptions) -> DataFrame:
```
Comment (nfx) [inline src/databricks/labs/remorph/reconcile/connectors/data_source.py:23]:
```suggestion
    def read_data(self, catalog: str, schema: str, query: str, options: JdbcReaderOptions) -> DataFrame:
```
nit: can you still rename all arguments to make them reasonably shorter? :)

File: src/databricks/labs/remorph/reconcile/query_config.py [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+
+    def get_table_name(self):
+        table_name = self.table_conf.source_name if self.layer == "source" else self.table_conf.target_name
+        if self.db_type == SourceType.ORACLE.value:
+            return "{{schema_name}}.{table_name}".format(  # pylint: disable=consider-using-f-string
+                table_name=table_name
+            )
+        return "{{catalog_name}}.{{schema_name}}.{table_name}".format(  # pylint: disable=consider-using-f-string
+            table_name=table_name
+        )
```
Comment (nfx) [inline src/databricks/labs/remorph/reconcile/query_config.py:53]:
```suggestion
        if self.db_type == SourceType.ORACLE.value:
            return f"{{schema_name}}.{table_name}"
        return f"{{catalog_name}}.{{schema_name}}.{table_name}"
```
# DO NOT disable `pylint` messages.

automating in https://github.com/databrickslabs/remorph/pull/191

**Review by nfx — CHANGES_REQUESTED** [review]
every time you self-review this code, look at the comments from codecov and make sure there's as little as possible of those:
image

File: src/databricks/labs/remorph/reconcile/connectors/data_source.py (hunk 12 above)
Comment (ravit-db) [reply src/databricks/labs/remorph/reconcile/connectors/data_source.py]: updated the arguments to shorter.

File: src/databricks/labs/remorph/reconcile/connectors/oracle.py [hunk 14]
```diff
@@ -16,16 +16,14 @@ def get_jdbc_url(self) -> str:
...
+            table_query = self._get_table_or_query(catalog, schema, query)
+            if jdbc_reader_options is None:
+                return self.reader(table_query).options(**self._get_timestamp_options()).load()
             return (
-                self.reader(query)
-                .options(
-                    **self._get_jdbc_reader_options(table_conf.jdbc_reader_options) | self._get_timestamp_options()
-                )
+                self.reader(table_query)
+                .options(**self._get_jdbc_reader_options(jdbc_reader_options) | self._get_timestamp_options())
```
[the remark this answers is not in the document]
Comment (ravit-db) [reply src/databricks/labs/remorph/reconcile/connectors/oracle.py]: resolved

File: src/databricks/labs/remorph/reconcile/query_config.py [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+class QueryConfig:
+    def __init__(self, table_conf: Table, schema: list[Schema], layer: str, db_type: str):
+        self.table_conf = table_conf
+        self.schema = schema
+        self.layer = layer
+        self.db_type = db_type
+        self.schema_dict = {v.column_name: v for v in schema}
+        self.tgt_column_mapping = table_conf.list_to_dict(ColumnMapping, "target_name")
+        self.src_column_mapping = table_conf.list_to_dict(ColumnMapping, "source_name")
+        self.transformations_dict = table_conf.list_to_dict(Transformation, "column_name")
```
[the remark this answers is not in the document]
Comment (ravit-db) [reply src/databricks/labs/remorph/reconcile/query_config.py]: updated.

File: src/databricks/labs/remorph/reconcile/query_config.py [hunk 16]
```diff
@@ -0,0 +1,69 @@
...
+
+    def get_threshold_columns(self):
```
[the remark this answers is not in the document]
Comment (ravit-db) [reply src/databricks/labs/remorph/reconcile/query_config.py]: resolved

File: src/databricks/labs/remorph/reconcile/query_config.py (hunk 13 above)
Comment (ravit-db) [reply src/databricks/labs/remorph/reconcile/query_config.py]: resolved

Comment (ravit-db) [reply src/databricks/labs/remorph/reconcile/query_config.py]: Fixed few and two pylint errors are un-resolved.one is an in-built function override and the other is with import error.

**Review by nfx — APPROVED** [review]
Lgtm

**Pull request merged**